### PR TITLE
Auth integration with AppCheck

### PR DESCRIPTION
**Pull request by aiwenisevan**
### Discussion

* Implement [ios sdk integration for AppCheck with Auth](https://b.corp.google.com/issues/265453331)

### Testing
* Passed all unit tests under FIRAuth
* Passed tests by using AppCheck Debug Provider

### API Changes

  * Check the [Auth&AppCheck Integration Guide](https://docs.google.com/document/d/1jNCMqUPXJLsGts1XFgo8tLqfBwEDkb6Y092pwVtUtkc/edit?resourcekey=0-n_Tdlw6x8RYpNftDd96BOQ#)

#no-changelog

### CODE REVIEW

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 1]
```diff
@@ -457,12 +460,13 @@ + (FIRAuth *)authWithApp:(FIRApp *)app {
   return (FIRAuth *)auth;
 }
 
-- (instancetype)initWithApp:(FIRApp *)app {
+- (instancetype)initWithApp:(FIRApp *)app appCheck:(nullable id<FIRAppCheckInterop>)appCheck {
```
Comment (renkelvin) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:463]: How about passing the appcheck directly to requesetConfiguration instead of modifying Auth initializer?

Comment (rosalyntan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:463]: Wouldn't it still have to go through the initializer? Since AppCheck is injected into FirebaseAuth via FirebaseComponents.

Comment (renkelvin) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:463]: Then I suggest pass it via FIRApp (similar to heartbeatLogger), so we can keep the initializer as simple as possible.

Comment (aiwenisevan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:463]: Thanks for the advice, But there will be more changes in other files like FIRApp and FIRAppInternal, if trying to pass it in FIRApp. I don't think we should add any logic into FIRApp.

**Comment by renkelvin** [comment]
Intuitively, I'd like to keep the FIRAuth initializer unchanged as it's the
main entrypoint of auth service. Let me see if there is a better workaround.

**Comment by rosalyntan** [comment]
> Intuitively, I'd like to keep the FIRAuth initializer unchanged as it's the
> main entrypoint of auth service. Let me see if there is a better workaround.

I took a look to see how RTDB does it -- @aiwenisevan you can probably do something like this: https://github.com/firebase/firebase-ios-sdk/blob/5f7d81cc19a1c8a98aa281626a372da39a0b5fcd/FirebaseDatabase/Sources/Api/FIRDatabaseComponent.m#L143

**Comment by aiwenisevan** [comment]
> > Intuitively, I'd like to keep the FIRAuth initializer unchanged as it's the
> > main entrypoint of auth service. Let me see if there is a better workaround.
> 
> I took a look to see how RTDB does it -- @aiwenisevan you can probably do something like this:
> 
> https://github.com/firebase/firebase-ios-sdk/blob/5f7d81cc19a1c8a98aa281626a372da39a0b5fcd/FirebaseDatabase/Sources/Api/FIRDatabaseComponent.m#L143

**Comment by aiwenisevan** [comment]
@rosalyntan Have updated the pr according to the instance Thx!

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 2]
```diff
@@ -2175,8 +2180,11 @@ - (BOOL)getUser:(FIRUser *_Nullable *)outUser error:(NSError *_Nullable *_Nullab
   FIRComponentCreationBlock authCreationBlock =
       ^id _Nullable(FIRComponentContainer *_Nonnull container, BOOL *_Nonnull isCacheable) {
     *isCacheable = YES;
+    // id<FIRAppCheckInterop> appCheckInterop = FIR_COMPONENT(FIRAppCheckInterop, container);
```
Comment (renkelvin) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:2183]: Please remove.

**Review by renkelvin — COMMENTED** [review]
Looks great, thanks!

**Review by renkelvin — APPROVED** [review]

File: FirebaseAuth/Sources/Auth/FIRAuth.m (hunk 2 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m]: Removed, thx!

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 3]
```diff
@@ -357,7 +358,8 @@ @interface FIRAuth () <FIRLibrary, FIRComponentLifecycleMaintainer>
 @property(nonatomic, strong, nullable) FIRAuthStoredUserManager *storedUserManager;
 
 /** @fn initWithApp:
-    @brief Creates a @c FIRAuth instance associated with the provided @c FIRApp instance.
+    @brief Creates a @c FIRAuth instance associated with the provided @c FIRApp instance and
+   AppCheck Interop.
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:362]: nit: You can revert the change to this comment, now that we're no longer passing the App Check interop into the initializer.

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.m [hunk 4]
```diff
@@ -641,7 +645,22 @@ + (NSMutableURLRequest *)requestWithURL:(NSURL *)URL
   if (languageCode.length) {
     [request setValue:languageCode forHTTPHeaderField:kFirebaseLocalHeader];
   }
-  return request;
+  if (requestConfiguration.appcheck) {
+    [requestConfiguration.appcheck
+        getTokenForcingRefresh:false
+                    completion:^(id<FIRAppCheckTokenResultInterop> _Nonnull tokenResult) {
+                      if (tokenResult.error) {
+                        FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Backend/FIRAuthBackend.m:653]: This error message and code needs to be updated (and also changed to `FIRLogWarning`).

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h [hunk 5]
```diff
@@ -41,6 +42,10 @@ NS_ASSUME_NONNULL_BEGIN
     @brief The heartbeat logger used to add heartbeats to the corresponding request's header.
  */
 @property(nonatomic, copy, nullable) id<FIRHeartbeatLoggerProtocol> heartbeatLogger;
+/** @property appcheck
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:45]: nit: Can you add one line of white-space in between the previous property and this one?

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 6]
```diff
@@ -462,7 +464,8 @@ - (instancetype)initWithApp:(FIRApp *)app {
   self = [self initWithAPIKey:app.options.APIKey
                       appName:app.name
                         appID:app.options.googleAppID
-              heartbeatLogger:app.heartbeatLogger];
+              heartbeatLogger:app.heartbeatLogger
+                     appcheck:FIR_COMPONENT(FIRAppCheckInterop, app.container)];
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:468]: nit: Here and in the rest of the PR, can you please change `appcheck` to `appCheck`?

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h [hunk 7]
```diff
@@ -77,6 +82,19 @@ NS_ASSUME_NONNULL_BEGIN
                         heartbeatLogger:(nullable id<FIRHeartbeatLoggerProtocol>)heartbeatLogger
     NS_DESIGNATED_INITIALIZER;
 
+/** @fn initWithAPIKey:appID:heartbeatLogger:
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:85]: nit: `initWithAPIKey:appID:heartbeatLogger:appCheck:`

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h [hunk 8]
```diff
@@ -77,6 +82,19 @@ NS_ASSUME_NONNULL_BEGIN
                         heartbeatLogger:(nullable id<FIRHeartbeatLoggerProtocol>)heartbeatLogger
     NS_DESIGNATED_INITIALIZER;
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:83]: This can be deleted (and any remaining calls to this should be updated to use the initializer with both heartbeatLogger and appCheck).

File: FirebaseAuth/Tests/Sample/Sample/ApplicationDelegate.m [hunk 9]
```diff
@@ -45,6 +47,11 @@ - (BOOL)application:(UIApplication *)application
   [[FIRConfiguration sharedInstance] setLoggerLevel:FIRLoggerLevelInfo];
 
   // Configure the default Firebase application:
+  FIRAppCheckDebugProviderFactory *providerFactory =
```
Comment (rosalyntan) [inline FirebaseAuth/Tests/Sample/Sample/ApplicationDelegate.m:50]: I would remove Sample app changes for now (and just leave it in your local workspace for testing).

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h (hunk 5 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:45]: Sure!

File: FirebaseAuth/Sources/Auth/FIRAuth.m (hunk 3 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:362]: Sure, thx, solved.

File: FirebaseAuth/Sources/Auth/FIRAuth.m (hunk 6 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:468]: yes, Have changed

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h (hunk 7 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:85]: Solved

File: FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h (hunk 8 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Backend/FIRAuthRequestConfiguration.h:83]: deleted

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.m (hunk 4 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Backend/FIRAuthBackend.m:653]: Have updated the code to a unique number,  where can I find the correct message?

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 10]
```diff
@@ -2330,10 +2334,10 @@ - (nullable FIRUser *)getStoredUserForAccessGroup:(NSString *_Nullable)accessGro
   } else {
 #if TARGET_OS_TV
     if (self.shareAuthStateAcrossDevices) {
-      FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
-                  @"Getting a stored user for a given access group is not supported "
-                  @"on tvOS when `shareAuthStateAcrossDevices` is set to `true` (#8878)."
-                  @"This case will return `nil`.");
+      FIRLogWarning(kFIRLoggerAuth, @"I-AUT000018",
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:2337]: Ah, I think this change was meant for line 653 of `FIRAuthBackend.m`.

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.m (hunk 4 above)
Comment (rosalyntan) [reply FirebaseAuth/Sources/Backend/FIRAuthBackend.m:653]: Looks like the code update was accidentally made in the wrong location! (I added a comment there as well)

For the error message, you can use: `@"Error getting App Check token; using placeholder token instead. Error: %@"`

File: FirebaseAuth/Tests/Sample/Sample/ApplicationDelegate.m (hunk 9 above)
Comment (aiwenisevan) [reply FirebaseAuth/Tests/Sample/Sample/ApplicationDelegate.m]: removed. thx!

File: FirebaseAuth/Tests/Sample/Podfile [hunk 11]
```diff
@@ -17,6 +17,7 @@ target 'AuthSample' do
   pod 'FBSDKLoginKit', '< 13'
   pod 'GoogleSignIn', '~> 6'
   pod 'GTMSessionFetcher/Core'
+  pod 'FirebaseAppCheck'
```
Comment (rosalyntan) [inline FirebaseAuth/Tests/Sample/Podfile:20]: You can remove this from the PR as well.

**Review by rosalyntan — COMMENTED** [review]
Looks good! Just a couple more minor comments.

File: FirebaseAuth/Sources/Auth/FIRAuth.m (hunk 10 above)
Comment (aiwenisevan) [reply FirebaseAuth/Sources/Auth/FIRAuth.m:2337]: got it thx

File: FirebaseAuth/Tests/Sample/Podfile (hunk 11 above)
Comment (aiwenisevan) [reply FirebaseAuth/Tests/Sample/Podfile]: fixed all

**Review by rosalyntan — APPROVED** [review]
